go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,336 @@
+		if err != nil {
+			return err
+		}
+
+		observer.Progress(fmt.Sprintf("Verifying blcok: %d ", i))
+		if err := verifyProofLiveTrie(ctx, trie, observer); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// VerifyProofLiveTrie verifies the consistency of witness proofs for a live trie.
+// It reads the trie for the head block and loads accounts and storage slots.
+// It extracts witness proofs for these accounts and its storage,
+// and checks that values in the proof and the database match.
+func VerifyProofLiveTrie(ctx context.Context, dir string, config MptConfig, observer VerificationObserver) error {
+	trie, err := OpenFileLiveTrie(dir, config, NodeCacheConfig{})
+	if err != nil {
+		return err
+	}
+
+	observer.StartVerification()
+	err = errors.Join(
+		verifyProofLiveTrie(ctx, trie, observer),
+		trie.Close())
+	observer.EndVerification(err)
+	return err
+}
+
+func verifyProofLiveTrie(ctx context.Context, trie *LiveTrie, observer VerificationObserver) error {
+	rootHash, hints, err := trie.UpdateHashes()
+	if hints != nil {
+		hints.Release()
+	}
+	if err != nil {
+		return err
+	}
+
+	observer.Progress("Collecting states... ")
+	// collect all accounts for this block
+	visitor := accountCollectingVisitor{accounts: accountsMap{}, ctx: ctx}
+	if err := trie.VisitTrie(&visitor); err != nil {
+		return err
+	}
+	if interrupt.IsCancelled(ctx) {
+		return interrupt.ErrCanceled
+	}
+
+	observer.Progress(fmt.Sprintf("Collecting and Verifying proof... \t Accounts: %d", len(visitor.accounts)))
+	var itemsDone, percentReported int
+	for addr, data := range visitor.accounts {
+		if interrupt.IsCancelled(ctx) {
+			return interrupt.ErrCanceled
+		}
+
+		// create proof for account only
+		proof, err := trie.CreateWitnessProof(addr)
+		if err != nil {
+			return err
+		}
+
+		// verify the proof
+		if err := verifyAccountProof(rootHash, proof, addr, data); err != nil {
+			return err
+		}
+
+		// create proof for storage by batches
+		for _, keys := range data.storageKeyRanges(10) {
+			if interrupt.IsCancelled(ctx) {
+				return interrupt.ErrCanceled
+			}
+
+			proof, err := trie.CreateWitnessProof(addr, keys...)
+			if err != nil {
+				return err
+			}
+
+			if err := verifyStorageProof(rootHash, proof, addr, keys, data.storage); err != nil {
+				return err
+			}
+
+			itemsDone += len(keys)
+			percent := float64(itemsDone) / float64(visitor.numOfKeys) * 100
+			if int(percent) >= percentReported {
+				observer.Progress(fmt.Sprintf("Verification progress: %.2f%%", percent))
+				percentReported = int(percent) + 10 // report step min 10%
+			}
+		}
+	}
+	// finally, that addresses and keys not present in the proof are empty
+	return verifyProofEmptyValues(ctx, trie, rootHash, maps.Keys(visitor.accounts), 1000, observer)
+}
+
+// verifyProofEmptyValues verifies that the witness proof for addresses and keys that lead to empty accounts
+// and empty slots, because they are not present in the MPT and in a generated proof.
```

Comment:
nit: this sentence misses some word(s) ...

Reply:
fixed